README.md
karma-selenium-grid-launcher
========================

A plugin for Karma that launches browsers on a remote [Selenium Grid].
Forked from [karma-webdriver-launcher].

Uses the [selenium-webdriver NPM module] behind the scenes.

[karma-webdriver-launcher]: https://github.com/karma-runner/karma-webdriver-launcher
[Selenium Grid]: https://www.seleniumhq.org/docs/07_selenium_grid.jsp
[selenium-webdriver NPM module]: https://www.npmjs.com/package/selenium-webdriver

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Usage](#usage)
- [Launcher-Specific Configuration Keys](#launcher-specific-configuration-keys)
  - [arguments](#arguments)
  - [base](#base)
  - [closeBeforeQuit](#closebeforequit)
  - [delayLaunch](#delaylaunch)
  - [extensions](#extensions)
  - [firefoxPreferences](#firefoxpreferences)
  - [gridUrl](#gridurl)
  - [heartbeatInterval](#heartbeatinterval)
  - [options](#options)
  - [promptOn (deprecated)](#prompton-deprecated)
  - [resetBeforeQuit](#resetbeforequit)
  - [suppressWarning](#suppresswarning)
  - [windowGeometry](#windowgeometry)
  - [x-ua-compatible](#x-ua-compatible)
- [Killing Tests Prematurely](#killing-tests-prematurely)
- [Browser Quirks and Workarounds](#browser-quirks-and-workarounds)
  - [Video Playback on macOS Safari](#video-playback-on-macos-safari)
  - [iOS 10 Safari](#ios-10-safari)
  - [iOS 12 Safari](#ios-12-safari)
- [Examples](#examples)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Usage

```bash
$ npm install --save-dev karma-selenium-grid-launcher
```

In your karma.conf.js file define some custom launchers that inherit from
`SeleniumGrid`, providing the URL of the Selenium Grid as `gridUrl`.

```js
module.exports = function(karma) {

  ...

    config.set({

      ...

      customLaunchers: {
        'ie11': {
          base: 'SeleniumGrid',
          gridUrl: 'http://some-host:4444/wd/hub',
          browserName: 'internet explorer',
          platform: 'windows',
          version: '11',
          'x-ua-compatible': 'IE=EmulateIE7',
          heartbeatInterval: 10000
        }
      },

      browsers: ['ie11'],

      ...

    });


```

Other than the launcher-specific configuration keys described below, all keys
are passed along to the grid as capabilities, giving you the flexibility to
do anything the corresponding Selenium WebDriver supports (see the Selenium
docs for more information).

Note that if you're trying to test on several browsers/platforms and you're
using the default Selenium matcher, you'll probably want to make use of the
undocumented `applicationName` capability, as the only platforms that
Selenium will match are `WINDOWS`, `MAC`, and `LINUX`.

You may also wish to use the [browser constants] exported by selenium-webdriver
to make sure you're using the `browserName` expected by selenium.

[browser constants]: https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/index_exports_Browser.html

## Launcher-Specific Configuration Keys

The following keys provided as part of a `customLauncher` block will *not* be
interpreted as a capability to send to the Grid; instead, they are processed
in some way by the launcher.

### arguments

Array of arguments that should be used when launching the browser binary. Note
that only Chrome, Firefox, and IE accept this key; an error will be thrown if
another `browserName` is used.

### base

This is obviously only used by Karma to tell it from which base to extend, but
it's included here to make it clear that `base` will not be passed along as a
capability.

### closeBeforeQuit

Call `driver.close()` before calling `driver.quit()`. Set this to `true` to
work around browsers/devices which don't implicitly close the active window
when testing finishes.

See the [Browser Quirks and Workarounds] section for which browsers/devices
may require this for proper functionality.

### delayLaunch

Time in ms to wait before launching a browser. Use this when a browser must
end up as the foreground window for tests to pass, such as testing video
playback on Safari.

### extensions

Array of extension paths that should be installed on the remote browser. Note
that only Chrome and Firefox accept this key; an error will be thrown if
another `browserName` is used.

### firefoxPreferences

Object mapping Firefox preference keys to their desired value. If you wish to
override one of the preferences passed to firefox by default, this will do
the trick.

### gridUrl

The URL of the Selenium Grid to use to launch browsers. If not provided, it
will default to `http://localhost:4444/wd/hub`.

### heartbeatInterval

Interval in ms to send a heartbeat to avoid the session being killed by
timeout. If not provided or set to `0`, no heartbeat will be performed.

### options

Object mapping a method name for the appropriate `Options` object of the
browser to the value with which it should be called. See the
[selenium-webdriver API docs] for more information.

### promptOn (deprecated)

Send and then close a prompt when an element with the given `id` is found in
the document.

This option is deprecated, and is probably not needed in any scenario.

### resetBeforeQuit

Have the browser load `about:blank` before calling `driver.quit()`. Set this
to `true` to work around browsers/devices which don't support closing the
active window when testing finishes.

See the [Browser Quirks and Workarounds] section for which browsers/devices
may require this for proper functionality.

### suppressWarning

A value of `true` will suppress the internet explorer warning; any other value
will cause the warning to be displayed.

### windowGeometry

Object to use for setting the geometry of the window. If any of the keys
`width`, `height`, `x`, or `y` are present in the object, they will be used to
modify just that aspect of the window geometry. Any other keys will be ignored.

### x-ua-compatible

The value to use for an `X-UA-Compatible` `meta` tag. If not provided, Karma
will not add the `meta` tag.

[selenium-webdriver API docs]: https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/
[Browser Quirks and Workarounds]: #browser-quirks-and-workarounds

## Killing Tests Prematurely

If you stop tests prematurely (e.g. by hitting `Ctrl+c`) soon after launching
the tests, it may take some time for the browsers/devices to fully close. To
keep things simple(ish), there are a series of promises during
startup/initialization. When you stop the requests determines where in this
series there will be a call to `driver.quit()`.

If the remote driver has returned a session, then you will have to wait for
the browser/device to load the karma page and perform its initialization
before `driver.quit()` will be called.

If you're using mobile devices, also keep in mind that building/deploying
the controller app and/or starting a simulator can take some time. Since
the driver can only be used after this has finished, you must wait for this
process to complete to ensure proper cleanup.

From the time you stop the tests until a particular browser has closed, the
message `Waiting for ${browser} to quit... (${time}s)` will be output to the
console every 10s.

## Browser Quirks and Workarounds

This is a list of known compatibility issues for certain browsers/devices,
and how to work around those issues.

If you encounter any browser quirks not listed here, please open a PR that
adds it, even if there is no known workaround.

### Video Playback on macOS Safari

macOS Safari requires that it is the foreground window to play back video. Use
the [`delayLaunch`](#delayLaunch) option to ensure that Safari launches after
any other browsers on the same machine.

Note also that the video element must have some portion within the viewport
being displayed by the browser. In your tests, ensure this is done.

### iOS 10 Safari

iOS 10 Safari does not implicitly close the current window when `driver.quit()`
is called, as specified by the WebDriver spec. This causes the browser to
connect twice during the next test run; the old window and the newly-launched
window. You should use the [`closeBeforeQuit`](#closeBeforeQuit) option to
ensure the window is closed at the end of each test run.

### iOS 12 Safari

iOS 12 Safari does not support closing the current window either through
`driver.close()` or `driver.quit()`. This causes the old window to connect
to the next run, before being disconnected (the new test run reuses the old
window, so the page effectively refreshes). You should use the
[`resetBeforeQuit`](#resetBeforeQuit) option to ensure the window is set to
`about:blank` at the end of each test run.

## Examples

For a more detailed example, see [the example file].

[the example file]: https://github.com/squarebracket/karma-selenium-grid-launcher/blob/master/examples/karma.conf.js


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.0] - 2019-07-26
### Added
- `closeBeforeQuit` launcher option
- `resetBeforeQuit` launcher option
- `Waiting for <browser> to quit...` message
- Killing Tests Prematurely section to README
- Browser Quirks and Workarounds section to README

### Changed
- Deprecated `promptOn` launcher option
- Major code refactor

### Fixed
- Quitting while launching browsers should now be handled properly
- Reliability is generally better

## [0.2.0] - 2019-03-29
### Added
- `promptOn` launcher option
- `delayLaunch` launcher option
- `windowGeometry` launcher option

### Changed
- Make things more generic and promisified and generally better

## [0.1.2] - 2019-01-15
### Added
- Send and then close alert to non-Safari browsers to try to avoid focus problems

### Changed
- Error out after a certain number of heartbeat errors

## [0.1.1] - 2018-11-05
### Fixed
- Set browser display name based on `customLaunchers` key

## 0.1.0 - 2018-11-05
### Added
- Initial Code

[0.3.0]: https://github.com/squarebracket/karma-selenium-grid-launcher/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/squarebracket/karma-selenium-grid-launcher/compare/0.1.2...0.2.0
[0.1.2]: https://github.com/squarebracket/karma-selenium-grid-launcher/compare/0.1.1...0.1.2
[0.1.1]: https://github.com/squarebracket/karma-selenium-grid-launcher/compare/0.1.0...0.1.1


examples/karma.conf.js
module.exports = function (karma) {


  const gridUrl = 'http://some.remote-host.com:4444/wd/hub';
  const firefoxPreferences = {
    // use OS cert store on windows
    'security.enterprise_roots.enabled': true
  };

  let customLaunchers = {
    'chrome-win7': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: 'chrome',
      applicationName: 'chrome/win7',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true
    },
    'chrome-beta-win7': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: 'chrome',
      applicationName: 'chrome/win7',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true,
      options: {
        setChromeBinaryPath: "C:\\Program Files (x86)\\Google\\Chrome Beta\\Application\\chrome.exe"
      }
    },
    'firefox-win7': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: 'firefox',
      applicationName: 'firefox/win7',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true,
      firefoxPreferences: firefoxPreferences,
      options: {
        setBinary: 'c:\\program files\\mozilla firefox\\firefox.exe'
      },
      arguments: [
        '-headless'
      ]
    },
    'chrome-beta-win81': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: 'chrome',
      applicationName: 'chrome/win81',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true,
      options: {
        setChromeBinaryPath: "C:\\Program Files (x86)\\Google\\Chrome Beta\\Application\\chrome.exe"
      }
    },
    'IE11-win10': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: 'internet explorer',
      applicationName: 'ie11/win10',
      suppressWarning: true,
      heartbeatInterval: 10000,
    },
    'edge-win10': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: 'MicrosoftEdge',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true
    },
    'chrome-mac1011': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: wd.Browser.CHROME,
      applicationName: 'chrome/mac1011',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true,
      someCustomCapability: 'customValue'
    },
    'chrome-beta-mac1011': {
      base: 'SeleniumGrid',
      gridUrl: gridUrl,
      browserName: wd.Browser.CHROME,
      applicationName: 'chrome/mac1011',
      heartbeatInterval: 10000,
      acceptInsecureCerts: true,
      options: {
        setChromeBinaryPath: "/Applications/Google Chrome Beta.app/Contents/MacOS/Google Chrome"
      }
    }
  };

  config.set({
    basePath: './',
    frameworks: ["jasmine"],
    reporters: ['progress'],
    logLevel: karma.LOG_INFO,
    customLaunchers: customLaunchers,
    browsers: ['firefox-win7', 'chrome-beta-win81', 'chrome-mac1011'];
    files: [
      "tests/*.spec.js"
    ],
    singleRun: true
  });
}


index.js
var wd = require('selenium-webdriver');
var urlModule = require('url');
var urlparse = urlModule.parse;
var urlformat = urlModule.format;
var chrome = require('selenium-webdriver/chrome');
var edge = require('selenium-webdriver/edge');
var firefox = require('selenium-webdriver/firefox');
var ie = require('selenium-webdriver/ie');
var safari = require('selenium-webdriver/safari');
var until = wd.until;

const CREATING_SESSION = 'CREATING_SESSION';
const CREATED_SESSION = 'CREATED_SESSION';
const WAITING = 'WAITING';
const ignoreArgs = ['base', 'gridUrl', 'suppressWarning', 'x-ua-compatible',
  'heartbeatInterval', 'promptOn', 'delayLaunch', 'windowGeometry'];
// default preferences shamelessly taken from karma-firefox-launcher
const defaultFirefoxPrefs = {
  'browser.shell.checkDefaultBrowser': false,
  'browser.bookmarks.restore_default_bookmarks': false,
  'dom.disable_open_during_load': false,
  'dom.max_script_run_time': 0,
  'extensions.autoDisableScopes': 0,
  'browser.tabs.remote.autostart': false,
  'browser.tabs.remote.autostart.2': false,
  'extensions.enabledScopes': 15,
};
// default chrome args shamelessly taken from karma-chrome-launcher
const defaultChromeArgs = [
  '--no-default-browser-check',
  '--no-first-run',
  '--disable-default-apps',
  '--disable-popup-blocking',
  '--disable-translate',
  '--disable-background-timer-throttling',
  // on macOS, disable-background-timer-throttling is not enough
  // and we need disable-renderer-backgrounding too
  // see https://github.com/karma-runner/karma-chrome-launcher/issues/123
  '--disable-renderer-backgrounding',
  '--disable-device-discovery-notifications',
];
const defaultIeOptions = {
  browserAttachTimeout: 30000,
};

var SeleniumGridInstance = function (name, args, logger, baseLauncherDecorator,
    captureTimeoutLauncherDecorator, retryLauncherDecorator) {
  if (!args.browserName) {
    throw new Error('browserName is required!');
  }

  var log = logger.create('SeleniumGrid');

  var gridUrl = args.gridUrl || 'http://localhost:4444/wd/hub';
  var self = this;

  // Intialize capabilities with default values
  const capabilities = new wd.Capabilities({
    platform: 'ANY',
    testName: 'Karma test',
    version: ''
  });

  if (args.browserName === 'internet explorer') {
    if (!args.suppressWarning) {
      log.warn('Internet Explorer requires some specific configuration to work properly. ' +
        'Follow the instructions on https://github.com/SeleniumHQ/selenium/wiki/InternetExplorerDriver#required-configuration ' +
        'and then add `suppressWarning` to the launcher config.');
    }
    // workaround until IE options are properly implemented in selenium-webdriver
    capabilities.set('se:ieOptions', defaultIeOptions);
  }

  const options = {};
  options[wd.Browser.CHROME] = new chrome.Options();
  options[wd.Browser.EDGE] = new edge.Options();
  options[wd.Browser.FIREFOX] = new firefox.Options();
  options[wd.Browser.IE] = new ie.Options();
  options[wd.Browser.SAFARI] = new safari.Options();

  Object.keys(defaultFirefoxPrefs).forEach((pref) => {
    options[wd.Browser.FIREFOX].setPreference(pref, defaultFirefoxPrefs[pref]);
  });
  options[wd.Browser.CHROME].addArguments(defaultChromeArgs);

  Object.keys(args).forEach(function (key) {
    if (ignoreArgs.indexOf(key) !== -1) {
      // used strictly for karma
      return;
    }

    if (key === 'firefoxPreferences') {
      Object.keys(args.firefoxPreferences).forEach((pref) => {
        options[wd.Browser.FIREFOX].setPreference(pref,
          args.firefoxPreferences[pref]);
      });
      return;
    }

    if ((key === 'arguments' || key === 'extensions' || key === 'options') &&
        (!options[args.browserName])) {
      throw new Error(key + ' not supported when using non-standard browser ' +
        args.browserName + '; you must use the equivalent capability');
    }

    if (key === 'arguments') {
      if (args[key].constructor !== Array) {
        throw new Error('arguments must be an Array for ' + args.browserName);
      }
      if (!options[args.browserName].addArguments) {
        throw new Error('arguments not supported for ' + args.browserName);
      }
      options[args.browserName].addArguments(args[key]);
      return;
    }

    if (key === 'extensions') {
      if (args[key].constructor !== Array) {
        throw new Error('extensions must be an Array for ' + args.browserName);
      }
      if (!options[args.browserName].addExtensions) {
        throw new Error('extensions not supported for ' + args.browserName);
      }
      options[args.browserName].addArguments(args[key]);
      return;
    }

    if (key === 'options') {
      if (!typeof args[key] === 'object') {
        throw new Error('options must be an object for ' + args.browserName);
      }
      Object.keys(args[key]).forEach((option) => {
        if (!options[args.browserName][option]) {
          throw new Error('option ' + option + ' not supported for ' + args.browserName);
        }
        options[args.browserName][option](args[key][option]);
      });
      return;
    }

    capabilities.set(key, args[key]);

  });

  baseLauncherDecorator(this);
  captureTimeoutLauncherDecorator(this);
  retryLauncherDecorator(this);

  self.name = name;
  self.browser = null;

  var heartbeatErrors = 0;
  var heartbeat;

  const heartbeatFunction = () => {
    log.debug('hearbeat for ' + self.name);
    self.browser.getTitle()
      .catch((err) => {
        heartbeatErrors++;
        log.error('Caught error for browser ' + self.name + ' during ' +
          'heartbeat: ' + err);
        if (err.name !== 'NoSuchSessionError') {
          log.error(err.stack);
        }
        if (heartbeatErrors >= 5) {
          log.error('Too many heartbeat errors, attempting to stop ' + self.name);
          args.heartbeatInterval && clearInterval(heartbeat);
          this.error = this.error || err;
          this.kill();
        }
      });
  };

  const promptFunction = (elId) => {
    return new Promise((resolve, reject) => {
      log.info('inside prompt function for ' + self.name);
      // sometimes we get stuck inside the prompt function when quitting,
      // so reject after a fixed amount of time just to unblock things
      let bumpTimeout = setTimeout(() => reject('prompt function stuck'), 10000);
      var windowRef;
      self.browser.getWindowHandle().then((ref) => {
        windowRef = ref;
        return self.browser.switchTo().frame(0);
      })
      .then(() => self.browser.wait(until.elementLocated(wd.By.id(elId))))
      .then(() => {
        log.debug('Trying to focus ' + self.name + ' with an alert...');
        return self.browser.switchTo().frame(null);
      })
      .then(() => self.browser.switchTo().window(windowRef))
      .then(() => self.browser.executeScript("alert('test')"))
      .then(() => self.browser.switchTo().alert())
      .then((alert) => alert.dismiss())
      .then(() => self.browser.switchTo().window(windowRef))
      .then(() => {
        resolve();
        clearTimeout(bumpTimeout);
      })
      .catch((err) => {
        log.error('caught in prompt for ' + self.name + ': ' + err);
        clearTimeout(bumpTimeout);
        reject(err);
      });
    });
  };

  // This is done by passing the option on the url, in response the Karma server will
  // set the following meta in the page.
  //   <meta http-equiv="X-UA-Compatible" content="[VALUE]"/>
  function handleXUaCompatible(urlObj) {
    if (args['x-ua-compatible']) {
      urlObj.query['x-ua-compatible'] = args['x-ua-compatible'];
    }
  }

  var loadTimeout = null;
  var started;
  var startPromise;

  this.awaitingKill = () => {
    return this.state === this.STATE_BEING_KILLED ||
      this.state === this.STATE_BEING_FORCE_KILLED;
  };

  this.on('start', (url) => {
    // reset kill state and error state so we don't auto-end the session immediately
    this.error = null;
    var urlObj = urlparse(url, true);

    handleXUaCompatible(urlObj);

    delete urlObj.search; //url.format does not want search attribute
    url = urlformat(urlObj);

    log.debug('Grid URL: ' + gridUrl);
    log.debug('Browser capabilities: ' + JSON.stringify(capabilities));

    let delayTime = 0;
    if (args.delayLaunch) {
      log.debug('Delaying launch of ' + args.browserName + ' for ' + args.delayLaunch + 'ms');
      this.state = WAITING;
      delayTime = args.delayLaunch;
    }

    loadTimeout = setTimeout(() => {
      let errorStage;
      log.debug('Launching ' + self.name);
      this.state = CREATING_SESSION;
      startPromise = new Promise((resolve, reject) => {
        self.browser = new wd.Builder()
          .setChromeOptions(options[wd.Browser.CHROME])
          .setEdgeOptions(options[wd.Browser.EDGE])
          .setFirefoxOptions(options[wd.Browser.FIREFOX])
          .setIeOptions(options[wd.Browser.IE])
          .setSafariOptions(options[wd.Browser.SAFARI])
          .usingServer(gridUrl)
          .withCapabilities(capabilities)
          .build()

        self.browser.then(() => {
          log.debug(self.name + ' started');
          resolve();
          started = true;
          this.state = CREATED_SESSION;
          startPromise = startPromise.then(() => {
            log.debug(self.name + ' resolved startPromise; navigating to karma page');
          });
          startPromise = startPromise.then(() => self.browser.get(url))
            .then(() => {
              log.debug(self.name + ' loaded karma; running applicable initialization');
              return Promise.resolve();
            })
            .catch(err => {
              errorStage = 'navigate';
              return Promise.reject(err);
            });
          if (args.windowGeometry) {
            startPromise = startPromise.then(() => {
              log.debug(self.name + ' setting window geometry');
              return self.browser.manage().window().setRect(args.windowGeometry).then(() => {
                log.debug(self.name + ' set window geometry');
                return Promise.resolve();
              });
            });
          }
          if (args.promptOn) {
            startPromise = startPromise.then(() => promptFunction(args.promptOn));
          }
          if (args.heartbeatInterval) {
            // TODO: This should maybe be before the `get` call?
            startPromise = startPromise.then(() => {
              log.debug(self.name + ' setting heartbeat');
              heartbeat = setInterval(heartbeatFunction, args.heartbeatInterval);
              return Promise.resolve();
            });
          }
          startPromise = startPromise.then(() => {
            log.debug(self.name + ' initialized');
            return Promise.resolve();
          });
          startPromise = startPromise.catch((err) => {
            if (errorStage) {
              // re-raise error caught during navigation
              return Promise.reject(err);
            } else {
              errorStage = 'init';
              return Promise.reject(err);
            }
          });
          return startPromise;
        })

        .catch((err) => {
          if (this.awaitingKill()) {
            log.debug('ignoring error from ' + self.name + '; browser is shutting down');
            reject(err);
            return Promise.resolve();
          } else {
            started = false;
            let message;
            if (errorStage === 'init') {
              message = self.name + ' encountered error during initialization: ' + err;
            } else if (errorStage === 'navigate') {
              message = self.name + ' encountered error navigating to karma page: ' + err;
            } else {
              message = self.name + ' was unable to create WebDriver session: ' + err;
            }
            log.error(message);
            this.error = err;
            reject(message);
            self._done();
          }
        }); // self.browser.then

      }); // startPromise

    }, delayTime);
  });

  let killInterval;
  let killElapsed = 0;

  this._stopSession = (done, startError) => {
    var self = this;
    let killPromise = Promise.resolve();

    clearInterval(heartbeat);

    if (!started) {
      log.info(self.name + ' not started... Killed ' + self.name);
      done();
      return killPromise;
    }

    if (args.resetBeforeQuit) {
      // load a blank page and wait before quitting. for browsers that don't
      // close the window before quitting. noticed on real iOS 12 devices.
      killPromise = killPromise.then(() => {
        log.debug('Resetting ' + self.name + ' and pausing.');
        return new Promise((resolve, reject) => {
          self.browser.get('about:blank').then(() => setTimeout(() => {
            log.debug('Reset ' + self.name + ' to about:blank');
            resolve();
          }, 10000));
        });
      });
    }

    if (args.closeBeforeQuit) {
      // explicitly call browser.close(), which should be unnecessary according
      // to webdriver. noticed on real iOS 10 devices.
      killPromise = killPromise.then(() => {
        log.debug('Closing browser window for ' + self.name + '.');
        return self.browser.close().then(() => {
          log.debug('Closed browser window for ' + self.name);
          this.state = 'CLOSED';
          return Promise.resolve();
        });
      });
    }

    killPromise = killPromise.then(() => {
      log.debug('Quitting WebDriver for ' + self.name + '.');
      return self.browser.quit()
        .then(() => {
          log.info('Killed ' + self.name + '.');
          done();
          return Promise.resolve();
        });
    });

    killPromise = killPromise.catch((err) => {
      if (err.name !== 'NoSuchSessionError' && err !== startError) {
        // ignore NoSuchSessionErrors when killing
        log.error('Error stopping browser ' + self.name + ': ' + err.toString());
      }
      done();
      return Promise.resolve();
    });

    return killPromise;
  };

  this.on('kill', (done) => {
    var self = this;
    log.info('Trying to kill ' + self.name);
    const end = () => {
      self._done();
      if (done) {
        done();
      }
    };
    const stopSession = (err) => {
      return new Promise((startPromiseResolve, startPromiseReject) => {
        this._stopSession(end, err).then(() => {
          clearInterval(killInterval);
          startPromiseResolve('shutting down');
          startPromiseReject('shutting down');
        });
      });
    };

    if (!self.browser) {
        log.info('Browser ' + self.name + ' has not yet launched.');
        loadTimeout && clearTimeout(loadTimeout);
        end();
      return Promise.resolve();
    }

    killInterval = setInterval(() => {
      killElapsed += 10;
      log.info('Waiting for ' + self.name + ' to quit... (' + killElapsed + 's)');
    }, 10000);

    return new Promise((resolve, reject) => {
      startPromise = startPromise.then(stopSession, stopSession);
    });
  });

};

// PUBLISH DI MODULE
module.exports = {
  'launcher:SeleniumGrid': ['type', SeleniumGridInstance]
};
